Clients: show did length estimation in downloadclient. Closes #3824

rcarpa opened the pull request:
Some refactoring was needed to make it possible. see individual commits for details. 
I would keep the commits separated on merge instead of squashing them.

Commit:
Clients: fix nrandom handling for downloading multiple items

When multiple items are provided to the downloadclient, they are
internally grouped to reduce the number of list_replicas calls.
Until now, nrandom was one of the fields used for grouping. This
means that two items with nrandom = 1 will be merged into one single
item. After that, list_replica is performed and one single file is
downloaded. I assume the desired behavior is to download one file
for each item, so disable grouping by the nrandom field.

Commit:
Clients: always list-dids in download client. #3824

The goal is to have length retrieved for each dids.

Commit:
Clients: rework downloadclient internals. #3824

- Reduce the number of internal dictionaries and their interaction.
- Separate the code needed to prepare list_replicas arguments and the
one preparing the options needed for download.
- Rework how resolved replicas are mapped back to input items:
until this commit, side effects where possible across input items.
For example, when two input items had different base_dir, some file
DIDs could have been downloaded to both dirs, even if other
filters (scheme, nrandom, rse_expression) should have avoided that.

Also, fully rely on serverside nrandom selection in downloadclient
The filtering on client side after doing it on serverside is redundant.

Commit:
Clients: show did length estimation in downloadclient. Closes #3824

mlassnig commented on lib/rucio/tests/test_download.py:
```diff
@@ -409,26 +512,45 @@ def test_trace_copy_out_and_checksum_validation(vo, rse_factory, did_factory, do
 
 
-def test_download_blocklisted_replicas(rse_factory, did_factory, download_client):
+def test_download_blacklisted_replicas(rse_factory, did_factory, download_client):
```
blocklist was quite alright :-)

mlassnig commented on bin/rucio:
```diff
@@ -1090,30 +1090,33 @@ def download(args):
-        print('{0:40} {1:6d}'.format('Total files: ', ds_total))
+
+        if length:
+            print('{0:40} {1:6d}'.format('Total files (did): ', length))
```
I would uppercase `DID`

mlassnig commented on bin/rucio:
```diff
@@ -1090,30 +1090,33 @@ def download(args):
+
+        if length:
+            print('{0:40} {1:6d}'.format('Total files (did): ', length))
+            print('{0:40} {1:6d}'.format('Total files (after filter):   ', ds_total))
```
shorter: `filtered`

rcarpa commented:
I cannot understand why that multi-vo tests fails after adding this function. But it's consistent :/

bari12 commented:
@Thysk any idea what is going on with the Multi-VO tests here? It is a very strange behaviour @rcarpa is noticing. Basically the code he changed does not touch the failing test at all. But it is working on master...

thysk commented:
@bari12 It looks like its succeeding in the 1st VO, but then is failing in the second VO
But given its the same process each time not really sure why
If this something that is failing in other PRs as well?
As there are multiple tests after this that function perfectly well, and perform almost identical tasks
Is there a way to see the full logs of the tests? because after the assert that fails there is another assert and I would be curious to see if that had the correct number of files (though it looks from the bytes that no files have been added to the dataset)

Given that all the exact same processes are used in the next test (test_account_counter_judge_evaluate_detach)
In am not sure how it is only failing the one test. And only on this PR, as, as you said, the PR doesnt touch any of the code that runs in the tests (as far as I can tell)

bari12 commented:
@rcarpa Please rebase and run again.
I don't think this is failing in other tests...

rcarpa commented:
I rebased and the problem persists :confused:

Commit:
Tests: cleanup updated dids before judge evaluator tests

Otherwise, the queue of updated dids can be bigger than 100 (default)
when evaluator will be executed. This will result in failing tests.

bari12 commented:
Thanks @rcarpa 
I'll bring this in as a feature for 1.27 though, since I want to stabilise the 1.26 clients a bit at the moment and not introduce yet another change.
